slapp/transfers/utils.py

```diff
@@ -1,109 +1,171 @@
+    """
+    responses = np.array(responses)
+    server_responses = np.array([
+            r['response']['ResponseMetadata']['HTTPStatusCode']
+            for r in responses])
+    success = responses[server_responses == 200]
+    failed = responses[server_responses != 200]
+    return success.tolist(), failed.tolist()
+
+
+def get_checksum(file_name: Union[pathlib.Path, str]) -> str:
+    """returns the base64-encoded 128-bit MD5 digest of the file contents
+
+    Parameters
+    ----------
+    file_name: path-like object
+       contents will be checksummed
 
     Returns
     -------
-    exists : bool
+    checksum: str
+        base64-encoded 128-bit MD5 digest
 
     """
-    exists = False
-    client = boto3.client('s3')
-    try:
-        client.head_object(Bucket=bucket, Key=key)
-        exists = True
-    except ClientError:
-        pass
-    return exists
-
-
-def local_s3_compare(file_name, bucket, key):
-    uri = s3_uri(bucket, key)
-    client = boto3.client('s3')
-
-    try:
-        assert object_exists(bucket, key)
-    except AssertionError:
-        raise S3TransferException(f"{uri} does not exist")
-    try:
-        tfile = tempfile.NamedTemporaryFile()
-        client.download_file(bucket, key, tfile.name)
-        assert filecmp.cmp(file_name, tfile.name)
-    except AssertionError:
-        raise S3TransferException(
-                f"destination object {uri} does not match "
-                f"source file {file_name}")
-    finally:
-        tfile.close()
-
-    return True
-
-
-def upload_file(file_name, bucket, key=None, skipcheck=False):
+    hash_object = hashlib.md5(open(file_name, "rb").read())
+    checksum = base64.b64encode(hash_object.digest()).decode('utf-8')
+    return checksum
+
+
+def upload_file(
+        client: Union[ConfiguredUploadClient, botocore.client.BaseClient],
+        file_name: Union[pathlib.Path, str],
+        bucket: str,
+        key: str) -> UploadResult:
     """Upload a file to an S3 bucket
 
     Parameters
     ----------
+    client: ConfiguredUploadClient
+        has a put_object() method
     file_name: path-like object
         full path to local file
     bucket: str
         name of bucket
     key: str
-        object key. If not passed, object key will be
-        basename of the file_name
-    skipcheck: bool
-        whether to skip the roundtrip check. Could be useful
-        in non-critical moments for large files
+        object key.
 
     Returns
     -------
-    uri : str
-        URI to S3 object just uploaded
+    result : UploadResult
+        contains the `file_name`, the destination `bucket` and `key`
+        and the server `response`
 
     """
-    if key is None:
-        key = pathlib.PurePath(file_name).name
+    checksum = get_checksum(file_name)
+    with open(file_name, 'rb') as fp:
+        response = client.put_object(
+                Body=fp,
+                Bucket=bucket,
+                Key=key,
+                ContentMD5=checksum)
```

Comment:
This is probably driving a big performance increase. I'm curious about the difference between running it in parallel vs. just removing the roundtrip from prior version.

Reply:
I suspect another difference was the first version created a client object per upload, whereas clients are being re-used now. For thousands of objects, I think this can be significant.